app/models/server/raw/ServerEvents.ts

```diff
@@ -0,0 +1,68 @@
+import { Collection, ObjectId } from 'mongodb';
+
+import { BaseRaw } from './BaseRaw';
+import { IServerEvent, IServerEventType } from '../../../../definition/IServerEvent';
+import { IServerEventsRepository } from '../definition/IServerEventsRepository';
+import { IUser } from '../../../../definition/IUser';
+
+export class ServerEventsRaw extends BaseRaw implements IServerEventsRepository {
+	public readonly col!: Collection<IServerEvent>;
+
+	async insertOne(data: Omit<IServerEvent, '_id'>): Promise<any> {
+		if (data.u) {
+			data.u = { _id: data.u._id, username: data.u.username } as IUser;
+		}
+		return this.col.insertOne({
+			_id: new ObjectId().toHexString(),
+			...data,
+		});
+	}
+
+	async findLastFailedAttemptByIp(ip: string): Promise<IServerEvent> {
+		return (await this.col.find({
+			ip,
+			t: IServerEventType.FAILED_LOGIN_ATTEMPT,
+		}, { sort: { ts: -1 }, limit: 1 }).toArray())[0];
+	}
+
+	async findLastFailedAttemptByUsername(username: string): Promise<IServerEvent> {
+		return (await this.col.find({
```

Comment:
any reason to now use `findOne`?

Reply:
@sampaiodiego it is just to enforce the sort. Do you think it is safe to use the findOne to get the last document?

Reply:
yes, it is safe =)